### PR TITLE
WIP: Generate requires for prepare from L1

### PR DESCRIPTION
**Pull request by pvalena**
to be able to install them.

_ _ _ _ 

**Not tested.**

### CODE REVIEW

File: tmt/steps/discover/fmf.py [hunk 1]
```diff
@@ -89,6 +89,14 @@ def tests(self):
         """ Return all discovered tests """
         return self._tests
 
+    def requires(self):
+        """ Return all tests' requires """[
+        requires = []
+        for test in tmt.Tree(testdir).tests(filters=self.filters)
+            if test.enabled
+                requires += test.require if test.require
```
Comment (pvalena) [inline tmt/steps/discover/fmf.py]: Maybe this should be `requires.append(...`.

@psss  Is the behaviour same?

**Comment by lukaszachy** [comment]
So, I've been playing with this and it seems to work for me. `result: 1 test passed, 0 tests failed`.

File: tmt/base.py [hunk 2]
```diff
@@ -220,6 +221,7 @@ def export(self, format_='yaml', keys=None, environment=None):
             data = dict()
             data['test'] = self.test
             data['path'] = self.path
+            data['require'] = " ".join(self.require)
```
Comment (pvalena) [inline tmt/base.py]: I think it should be a list :).

It gets processsed by `Prepare` which expects a list of packages to install...

File: tmt/steps/prepare/__init__.py [hunk 3]
```diff
@@ -53,8 +53,13 @@ def go(self):
 
         packages = []
         for step in self.plan.steps():
-            if getattr(step, 'requires', False):
-                packages += step.requires()
+            if hasattr(step, 'steps'):
+                for inner_step in step.steps:
+                    if getattr(inner_step, 'requires', False):
+                        packages += inner_step.requires()
```
Comment (pvalena) [inline tmt/steps/prepare/__init__.py]: That's strange. This shouldn't be the case... let me test+fix in #159, please.

**Comment by pvalena** [comment]
@lukaszachy thanks for debugging and filling in the gabs!

The ident is ... a github thing, sadly. I'll do it the usual way next time.

Please let me fix the functionality of Prepare in its PR (#159), as I'm already modifying it.

**Comment by lukaszachy** [comment]
> Please let me fix the functionality of Prepare in its PR (#159), as I'm already modifying it.
Sure think, I'll try it once ready - I am not blocked as locally it works for me even today.

File: tmt/steps/prepare/__init__.py (hunk 3 above)
Comment (pvalena) [reply tmt/steps/prepare/__init__.py]: @psss any idea why do we need `for inner_step in step.steps:`?

File: tmt/base.py (hunk 2 above)
Comment (psss) [reply tmt/base.py]: This is part of the export for `run.sh`. I don't think `run.sh` need to know about `require`. Or does it, @pvalena?

File: tmt/steps/prepare/__init__.py (hunk 3 above)
Comment (psss) [reply tmt/steps/prepare/__init__.py]: I think the `requires` method should be called only for `discover` (gathered L1 require) and `execute` (test framework execution requires like `beakerlib`) steps without touching inner data.

File: tmt/base.py (hunk 2 above)
Comment (pvalena) [reply tmt/base.py]: I have no Idea on this one, as @lukaszachy 's added this.

File: tmt/steps/prepare/__init__.py (hunk 3 above)
Comment (pvalena) [reply tmt/steps/prepare/__init__.py]: I agree, @lukaszachy why is the original implementation an issue for you?

**Comment by psss** [comment]
Gathering requires from L1 has been implemented in #218. Closing.

**Pull request closed without merging**